docs/code conventions (#905)
scania-digital-design-system/sdds: CODECONVENTIONS.md

# Tegel Design System - Code conventions

Welcome to the code conventions section of Tegel, a comprehensive design system that aims to provide a unified and consistent visual language for all of your digital products. In this section, we'll cover the coding standards and conventions that are used to build and maintain Tegel components.

## General
 - Use camelCase for variables and functions.
 - Use PascalCase for classes and interfaces.
 - Use template literals instead of concatenation.
 - Use meaningful variable and function names: Name your variables and functions in a way that clearly communicates their purpose.

> Avoid using single-letter variable names, index stead of `i` for index, be descriptive and name the variable `index`.

Reviewer: Should "stead" be "instead"? :)

> Avoid using single-letter variable names, index stead of `i` for index, be descriptive and name the variable `index`.

Author: spell fix

 - Use single quotes (') for strings.

## Component structure

> Our component files (e.g. accordion.tsx) should all follow the following structure:

Reviewer 2: ```suggestion
Our component files (e.g. accordion.tsx) should follow the following structure, in order:
```

1. @Component decorator with the component specific arguments.
2. Host element for the component (@Element).
3. Component props.
4. Component state.

> 5. Component variables.

Reviewer 2: ```suggestion
5. Private variables.
```

6. Public methods that are exposed on the component (@Method)
7. Events that are emitted from the component (@Event). 
8. Event listeners (@Listen)
9. Lifecycle methods (connectedCallback, componentWillLoad, componentDidRender, etc)

> 10. Methods that are used within the component.

Reviewer 2: ```suggestion
10. Private methods.
```

> 9. Lifecycle methods (connectedCallback, componentWillLoad, componentDidRender, etc)
> 10. Methods that are used within the component.

Reviewer 2: ```suggestion
9. Methods that are used within the component.
10. Lifecycle methods (connectedCallback, componentWillLoad, componentDidRender, etc)
```
What do you think about switching 9 and 10, since private methods might be used inside lifecycle methods, but not the other way around.

Author: True!

> 11. Render method.

Reviewer 2: WDYT about this structure instead? 
Thought it might be easier to remember if you can group them into three more general categories.

1. Event emitters and listeners.
   1. Event emitters (@Event).
   1. Event listeners (@Listen).
1. Props, state, and variables. 
   1. Host element for the component (@Element).
   1. Component props.
   1. Component state.
   1. Private variables.
1. Methods
   1. Prop and state and watchers (@Watch).
   1. Public methods that are exposed on the component (@Method)
   1. Private methods.
   1. Lifecycle methods
   1. Render method

Author: I do not think events should be first. I agree on grouping them in three groups but I would suggest switching the order of Event and Props?

Reviewer 2: I'm ok with switching 1 and 2

  @Prop() prop: string;
 
 /** Comment explaning  the use of the the prop */
  @Prop() secondProp: string;

  @State() state: boolean;

  variable: string

  /** Comment explaning the method. */
  @Method()
  async handleComponent(){
  }

  /** Comment explaning the event. */
  @Event({
    eventName: 'sddsEvent',
    composed: true ,
    cancelable: true,
    bubbles: true,
  })
  sddsEvent: EventEmitter<{}>;

  @Listen('sddsEvent', { target: 'body' })
  handleListener(){}

  handleClick(){}

  connectedCallback(){
  }

  componentWillLoad() {
  }

  render() {
    return (
      <Host>
      </Host>
    );
  }
}

```

## Events
The tegel components emit custom events to allow the users to repond to changes/updates in the components. These are all named using the 
sdds-prefix. This is done in order to not have conflicting events and to make it clear to the user the specified event is something that is emitted
from a tegel component.

The events are named according to our naming convention: 'sdds' + event.
For a click event this would result in the event being called `sddsClick`. To listen for these events in vanilla JS the event name

> should be passed to the `addEventListener` function as the first argument:

Reviewer 2: I don't think the rest of the events section after here fits under "coding conventions".
I think its more user documentation for how to consume events. Something that could fit on the storybook site, and later on the tegel website perhaps.

- import/no-extraneous-dependencies: "off"
  - This rule disallows the import of external modules that are not declared in the package.json file's dependencies, devDependencies, or optionalDependencies object. However, it is turned off in this configuration.
- import/prefer-default-export: "off"
  - This rule encourages the use of a default export when only one thing is being exported from a module. However, it is turned off in this configuration.
- no-underscore-dangle: "warn"
  - This rule disallows the use of underscores in variable and method names, except in specific situations. It is set to warn, meaning it will not cause the build to fail but will still generate a warning.
- import/no-unresolved: "warn"
  - This rule disallows importing modules that cannot be resolved by the specified paths. It is set to warn, meaning it will not cause the build to fail but will still generate a warning.
- no-return-assign: "warn"
  - This rule disallows assignments in `return` statements. It is set to warn, meaning it will not cause the build to fail but will still generate a warning.
- class-methods-use-this: "warn"

>   - This rule requires the use of `this` in class methods that use the class instance. It is set to warn, meaning it will not cause the build to fail but will still generate a warning.

Reviewer: Up until this point there has been a sentence after the explanation of the rule stating if the rule is set to warn or turned off, like "It is set to warn, meaning it will not cause the build to fail but will still generate a warning." or "However, it is turned off in this configuration.". After this rule it is not present anymore. Should we maybe decide on one approach and stick with it for the sake of consistency? :)

- "no-restricted-exports": "warn"
  - This rule is used to specify certain exports that should be avoided, and warns when they are used. For example, it can be used to disallow the use of certain functions or variables that are considered unsafe or deprecated.
- "radix": ["error", "as-needed"]
  - This rule requires that the `parseInt()` function always be called with a radix parameter, to avoid unexpected behavior when parsing strings to numbers. The `"as-needed"` option allows `parseInt()` to be called without a radix parameter only when the number is already in base 10.
- "no-use-before-define": "off"
  - This rule disallows the use of variables before they are defined, which can help catch common programming errors. However, it can also be overly strict in some cases, especially with function declarations. This option turns off the rule entirely.
- "no-console": "off"
  - This rule disallows the use of `console.log()` and other similar functions, which can help prevent debugging information from being included in production code. However, it can be useful during development and testing, so this option turns off the rule entirely.
- "no-undef": "off"

>   - This rule warns when a variable or function is used without being defined, which can help catch typos and

Reviewer: Looks like the sentence might have been cut off here? :)
